---
title: "scRNA-seq Processing, QC, and Normalization"
output:
  html_document:
    df_print: paged
  # html_notebook: default
  # pdf_document: default
editor_options: 
  chunk_output_type: console
---
# Description
The data we'll be using is a public dataset of ~8k PBMCs from 10x Genomics.

Sequencing data from scRNA-seq libraries is typically aligned to the transcriptome using the tool CellRanger (made by 10x Genomics). This process takes hours-to-days to run and requires access to a linux cluster, but fortunately, it's fairly straightforward so we'll just start working with the output generated by CellRanger. A few other tools have been developed to replace CellRanger, but for the most part, they just present improve run time rather than improved alignment/quantification.

*DISCLAIMER:* For this workshop, I will present a fairly standard workflow that is incredibly user-friendly. However, alternative tools exist for every step of the pipeline. Often, the tooling choice will largely just be a preference, but in some cases, these alternatives are demonstratably better. If you are considering getting more involved in the field, I encourage you to read the most recent benchmark papers and newest preprints--the field is still moving incredibly fast.

# Load package dependencies
```{r setup, include = FALSE}
knitr::opts_chunk$set(
  library(Seurat),
  library(tidyverse),
  library(datapasta),
  # library(GEOquery),
  # library(archive),
  library(DropletUtils),
  library(scuttle),
  # library(tensorflow),
  # library(cellassign), 
  
  library(org.Hs.eg.db),
  
  # automated cell type annotation ('label transfer') using reference data,
  # library(SingleR), 
  
  # a large collection of reference expression datasets with curated cell type labels for use with SingleR package
  # library(celldex), 
  
  # quality control and visualization for scRNA-seq data of object class SingleCellExperiment
  library(scater),
  library(DESeq2)
  )
```

# Load the data into R
Seurat has an easy function to load in data generated by CellRanger. It requires the presence of three files that are outputed: barcodes.tsv, features.tsv, matrix.mtx.

This funciton will read those files into a single sparse matrix variable in R with genes as rownames and cell barcodes at colnames.

```{r eval=FALSE, include=FALSE}
pbmc.data <- Read10X(data.dir = "../data/filtered_gene_bc_matrices_8kpbmc/GRCh38/")
```

```{r eval=FALSE, include=FALSE}
#getGEOSuppFiles("GSE150290")

archive_extract(archive = "GSE150290_RAW.tar", dir = "GSE150290")

for (i in seq_along(
  list.files("GSE150290", pattern = ".tar.gz$")
  )
     ) {
  archive_extract(
    list.files("GSE150290", pattern = ".tar.gz$", full.names = TRUE)[i], 
    dir = list.files("GSE150290", pattern = ".tar.gz$",  full.names = TRUE) |> str_match("GSE150290.*_Pat\\d+-\\w?") |> pluck(i)
  )
  }

#https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3660646
# Sequenced reads were then mapped to GRCh38 whole genome using 10X Genomics' Cell Ranger 2.0 software's cellranger count function
# Cell Ranger output of raw_gene_bc_matrices were utilized for further downstream analysis in R with the package Seurat 2.3.4, only cells expressing at least 200 genes were kept for downstream analysis. Further filtration removed cells expressing greater than 50,000 UMI or 35% mitochondrial genes.
```

If working with public data, the names of files are often modified (eg. by NCBI's GEO system) and you will have to switch them back.

Alternatively, many papers will simply provide this genes-by-cells matrix as a CSV file. If that is the case, you can use something the following code to read it into R:
my_data <- read.csv("~/Folder/location/of/their_data.csv", row.names=1)

# Move the data into a "Seurat Object"
If you look in your environment, you'll see that the expression matrix has been loaded into R as the variable "pbmc.data". This is great, but once you start doing more analysis, it can start getting messy if you start making dozens of other variables containing things like cell metadata, UMAP coordinates, etc.

This issue is fixed by a special class of variable called a **Seurat object**. A seurat object serves essentially as an organized container to hold almost all types of data you will work with in scRNA-seq analysis. The two benefits of this are 1) it simply keeps everything more organized than managing dozens of independent R variables, and 2) it imposes "rules" that help prevent you from messing your data up (for example, it demands that the cells in your metadata table are an exact match with the cells in the expression matrix).

Let's store the expression matrix in a brand new Seurat object. I will generically name the seurat object "seurat" so that I can re-use code between experiments as much as possible

```{r eval=FALSE, include=FALSE}
seurat <- CreateSeuratObject(counts = pbmc.data, 
                             min.cells = 3, 
                             project = "PBMC")
```

```{r eval=FALSE, include=FALSE}
seurat
```

# Quality control & Filtering
Here, we will explore three main quality control metrics:
1) the percentage of transcripts per cell that align to the mitrochondrial genome
2) the number of genes with detectable transcripts in each cell
3) the number of unique transcriptions (based on UMIs) per cell

NOTE: The distribution of percent.mito is suuuuper sample-specific. You'll have to change the cutoffs for your own sample
```{r eval=FALSE, include=FALSE}
seurat[["percent.mito"]] <- PercentageFeatureSet(seurat, pattern = "^MT-")
VlnPlot(seurat1, 
        features = c("nFeature_originalexp", "nCount_originalexp", "subsets_Mito_percent"), 
        pt.size = 0.25,
        ncol = 3)
```

Alternatively, you can grab these values out of the Seurat object and plot them yourself:
```{r eval=FALSE, include=FALSE}
hist(seurat@meta.data$percent.mito, breaks=50)
```

You can also see that, consistent with the loss of cytoplasmic integrity, cells with high mitochondrial reads have low gene content:

```{r eval=FALSE, include=FALSE}
qplot(x=nFeature_RNA, y=percent.mito, data=seurat@meta.data)
```

Filtering thresholds are quite arbitrary right now, but here, removing cells with >7.5% mitochrondrial reads seems good. It's also common to remove cells with less than ~200 genes to remove "low information" cells. No cells in this dataset have fewer than that, but I'll still include it in the code so you can see how it is done

```{r eval=FALSE, include=FALSE}
seurat <- subset(seurat, 
               subset = nFeature_RNA > 200 & 
                 percent.mito < 7.5)

c(224, 257, 334.5, 336, 341, 346, 353, 363, 369, 398, 423.5, 425, 428, 438, 438, 455.5, 460, 468, 473, 484, 489, 504, 520, 522, 526, 528, 529, 530.5, 542, 546, 549, 550, 552, 595, 598, 607.5, 611, 612, 624, 640, 656, 658, 660, 664.5, 674, 681, 697, 730, 797, 811, 844, 888) |> hist()
```

```{r eval=FALSE, include=FALSE}
patients <- paste0(
  "GSE150290/", 
  list.files("GSE150290", recursive = TRUE) |> 
    str_extract(".*/hg19") |> 
    na.exclude() |>  
    unique()
  )

results <- vector("list",length = length(patients))
is.mito  <- read_delim("GSE150290/GSM4546300_Pat01-A/raw_gene_bc_matrices/hg19/genes.tsv", 
    delim = "\t", escape_double = FALSE, 
    col_names = FALSE, trim_ws = TRUE) |>  pull(2) |> str_which("MT-")

for (i in seq_along(patients))
  {
  results[[i]] <- read10xCounts(patients[i])
  results[[i]] <- results[[i]][,which(emptyDrops(counts(results[[i]]))$FDR <= 0.001)]
  results[[i]] <- quickPerCellQC(results[[i]], subsets=list(Mito=is.mito), sub.fields="subsets_Mito_percent")
  results[[i]] <- as.Seurat(results[[i]], data = NULL)
  # results[[i]][["percent.mt"]] <- PercentageFeatureSet(object = results[[i]], pattern = "^MT-")
  # results[[i]][["percent.hb"]] <- PercentageFeatureSet(object = results[[i]], pattern = "^HB[^(P)]")
  # # Cho et al also removed cells with (1) sd of all genes per cell < 1; (2) zero UMI count for 90% of all genes
  # results[[i]] <- subset(results[[i]], subset = 
  #                          nCount_RNA < 20000 & #sum of counts aka library size
  #                          nCount_RNA > 100 & #sum of counts aka library size
  #                          nFeature_RNA > 1000 & #number of detected features
  #                          percent.mt < 10 &
  #                          percent.hb < 10 )
  results[[i]] <- NormalizeData(object = results[[i]])
  results[[i]] <- FindVariableFeatures(object = results[[i]])
  }

# saveRDS(results, file="./output/results.rds")

# select integration features
features <- SelectIntegrationFeatures(object.list = results)

# run PCA on each object in the list, which is required for running the alternative reciprocal PCA workflow
results <- lapply(X = results, FUN = function(x) {
    x <- ScaleData(x, features = features, verbose = FALSE)
    x <- RunPCA(x, features = features, verbose = FALSE)
}) 

full.features <- results[[1]] |> rownames()

# find integration anchors (CCA)
# anchors <- FindIntegrationAnchors(object.list = results,
#                        anchor.features = features)

# reciprocal PCA 
anchors <- FindIntegrationAnchors(object.list = results, 
                                  reference = c(1,3), 
                                  reduction = "rpca",
                                  dims = 1:50)

# saveRDS(anchors, file= "./output/anchors.rds")

# integrate data
seurat.integrated <- IntegrateData(anchorset = anchors, dims = 1:50, features.to.integrate = full.features)

seurat.integrated <- ScaleData(object = seurat.integrated)
seurat.integrated <- RunPCA(object = seurat.integrated)
seurat.integrated <- RunUMAP(object = seurat.integrated, dims = 1:50)

p3 <- DimPlot(seurat.integrated, reduction = 'umap', group.by = 'Sample') + theme(legend.position = "none")

p4 <- DimPlot(seurat.integrated, reduction = 'umap', group.by = 'Type',
              cols = c('red','green','blue'))
```

```{r annotation1, eval=FALSE, include=FALSE}
library(tensorflow)
# need to install tensorflow R package first (above)
# then run tensorflow::install_tensorflow(extra_packages='tensorflow-probability'), 
# then install cellassign from github: https://github.com/irrationone/cellassign
library(cellassign) 

# create a list of markers
# you can find cell specific markers here: http://biocc.hrbmu.edu.cn/CellMarker/
not_cd45_marker_list <- list(
  endothelial = c("PLVAP", "KDR", "PTPRB"),
  enteroendocrine = c("CHGA", "GAST", "PROX1"),
  fibroblast = c("MMP2", "PDGFRA", "MYL9", "FN1", "CAV1"),
  gmc = c("MUC6", "TFF2"),
  metaplasia = c("TFF3", "CDX1", "CDX2"),
  chief = c("PGC"),
  goblet = c("MUC2", "ITLN1", "HES6"),
  metaplastic.stem = c("OLFM4", "REG1A", "CLDN3"),
  proliferative = c("CDKN2A", "MKI67", "RBP4"),
  pmc = c("GKN1", "GKN2", "MUC5AC"),
  tumor = c("EPCAM", "CDH17", "COL3A1", "PDGFRB")
  )

# convert your marker gene list from above to a matrix
not_cd45_marker_matrix <- marker_list_to_mat(not_cd45_marker_list, include_other = FALSE)

# make sure all your markers were actually observed in your single cell data.  Remove markers that were not detected
marker_in_sce <- match(rownames(not_cd45_marker_matrix), rowData(seurat.sce)$Symbol)
# stopifnot(all(!is.na(marker_in_sce)))

not_cd45_marker_matrix <- not_cd45_marker_matrix |>  cbind(marker_in_sce)  |> na.omit()

marker_in_sce <- marker_in_sce |> na.omit()

#subset data to include only markers
sce_marker <- seurat.sce[marker_in_sce, ]
stopifnot(all.equal(rownames(not_cd45_marker_matrix), rowData(sce_marker)$Symbol))

# compute size factors

# assay(seurat.sce, "counts") <- 2^(assay(seurat.sce))
assay(sce_marker, "counts") <- 2^(assay(sce_marker))

# seurat.sce <- scran::computeSumFactors(seurat.sce) # this step takes two hours

# saveRDS(seurat.sce, file="./output/seurat.sce.rds")
# saveRDS(sce_marker, file="./output/sce_marker.rds")

sf <- edgeR::calcNormFactors(counts(seurat.sce))

# run cellAssign
fit <- cellassign(
  exprs_obj = sce_marker,
  marker_gene_info = not_cd45_marker_matrix,
  s = sf,
  shrinkage = TRUE,
  max_iter_adam = 50,
  min_delta = 2,
  verbose = TRUE)
# a newer version of TensorFlow is required than available via devtools
# Issue created for developer https://github.com/Irrationone/cellassign/issues/98
```

```{r annotation2}
# library(org.Hs.eg.db)
# 
# library(SingleR) #automated cell type annotation ('label transfer') using reference data
# library(celldex) #a large collection of reference expression datasets with curated cell type labels for use with SingleR package
# 
# library(scater) #quality control and visualization for scRNA-seq data

# seurat_integrated_two_ref <- readRDS("~/workshop/output/seurat_integrated_two_ref.rds")
# seurat.sce <- as.SingleCellExperiment(seurat_integrated_two_ref)

seurat.sce <- readRDS("~/workshop/output/seurat.sce.rds")

annots <- AnnotationDbi::select(org.Hs.eg.db, keys=rownames(seurat.sce),
columns=c("SYMBOL","GENENAME"), keytype="ENSEMBL") |>  
  # filter(!is.na(SYMBOL)) |> # remove rows that don't have a symbol
  dplyr::add_count(ENSEMBL) |> filter(n == 1) |>  # remove rows that map to two symbols
  dplyr::select(-n)

rowData(seurat.sce)$Symbol <- rownames(seurat.sce) |> as.data.frame() |> dplyr::rename(ENSEMBL = 1) |> left_join(annots) |>  dplyr::rename(ensembl = ENSEMBL, Symbol = SYMBOL, gene_name = GENENAME) |> dplyr::select(Symbol) 

# seurat.sce <- seurat.sce[!is.na(rowData(seurat.sce)$Symbol), ]

rownames(seurat.sce) <- rowData(seurat.sce)$Symbol

Monaco.data <- MonacoImmuneData(ensembl = FALSE) #114 bulk RNA-seq samples of sorted immune cell populations that can be found in GSE107011.

predictions <- SingleR(test=seurat.sce, assay.type.test=1, 
                       ref=Monaco.data, labels=Monaco.data$label.main)

plotScoreHeatmap(predictions)

# saveRDS(predictions, file="./output/predictions.rds")

#now add back to singleCellExperiment object (or Seurat objects)
seurat.sce[["SingleR.labels"]] <- predictions$labels |> as.factor()
# plotUMAP(seurat.sce, colour_by = "SingleR.labels")
```

```{r metadata}
# patient metadata columns taken from the supplementary file
patient_data <- data.frame(
  stringsAsFactors = FALSE,
           patient = c("01", "01", "02", "02", "03", "03", "04", "04", "05", "05", "06", "06", "07", "07", "08", "08", "09", "09", "10", "10", "11", "11", "12", "12", "13", "13", "14", "14", "15", "15", "16", "16", "17", "17", "18", "18", "19", "19", "20", "20", "21", "22", "22", "23", "23", "24", "24", "25", "26", "27", "28", "29") |> as.factor(),
            gender = c("M","M","M","M","F","F",
                       "F","F","M","M","M","M","M","M","F","F","M","M",
                       "M","M","F","F","F","F","M","M","F","F","M",
                       "M","M","M","M","M","M","M","M","M","F","F",
                       "M","M","M","M","M","F","F","F","F","F","F","F") |> as.factor(),
               age = c(75L,75L,72L,72L,66L,66L,
                       56L,56L,45L,45L,58L,58L,76L,76L,56L,56L,55L,55L,
                       54L,54L,55L,55L,62L,62L,70L,70L,45L,45L,66L,
                       66L,57L,57L,51L,51L,58L,58L,80L,80L,41L,41L,
                       63L,46L,46L,76L,76L,44L,44L,72L,59L,61L,48L,43L),
         diagnosis = c("EGC","EGC","AGC","AGC",
                       "EGC","EGC","EGC","EGC","EGC","EGC","EGC","EGC",
                       "EGC","EGC","EGC","EGC","EGC","EGC","EGC","EGC","AGC",
                       "AGC","AGC","AGC","EGC","EGC","AGC","AGC","AGC",
                       "AGC","EGC","EGC","EGC","EGC","EGC","EGC","EGC",
                       "EGC","AGC","AGC","AGC","EGC","EGC","AGC","AGC",
                       "AGC","AGC","CAG with IM","CSG","CSG","CSG",
                       "CAG with IM") |> as.factor(),
            lauren = c("Intestinal","Intestinal",
                       "Diffuse","Diffuse","Diffuse","Diffuse","Diffuse",
                       "Diffuse","Diffuse","Diffuse","Diffuse","Diffuse",
                       "Intestinal","Intestinal","Diffuse","Diffuse","Intestinal",
                       "Intestinal","Diffuse","Diffuse","Diffuse","Diffuse",
                       "Diffuse","Diffuse","Intestinal","Intestinal",
                       "Diffuse","Diffuse","Diffuse","Diffuse","Diffuse",
                       "Diffuse","Intestinal","Intestinal","Intestinal","Intestinal",
                       "Intestinal","Intestinal","Diffuse","Diffuse",
                       "Intestinal","Intestinal","Intestinal","Diffuse","Diffuse",
                       "Diffuse","Diffuse",NA,NA,NA,NA,NA) |> as.factor(),
           atrophy = c("N","N","N","N","Y","Y",
                       "Y","Y","N","N","N","N","Y","Y","N","N","Y","Y",
                       "N","N","Y","Y","N","N","N","N","N","N","N",
                       "N","N","N","N","N","N","N","N","N","N","N",
                       "Y","Y","Y","Y","Y","N","N",NA,NA,NA,NA,NA) |> as.factor(),
          location = c("LB, AW","LB, AW","LB, GC",
                       "LB, GC","Antrum, GC, PW","Antrum, GC, PW","Antrum, PW",
                       "Antrum, PW","MB, LC","MB, LC","UB, PW","UB, PW",
                       "Angle, PW","Angle, PW","LB, LC","LB, LC","MB, LC",
                       "MB, LC","LB, LC","LB, LC","MB, LC","MB, LC","LB, LC",
                       "LB, LC","Angle-antum, AW","Angle-antum, AW",
                       "Antrum, PW","Antrum, PW","MB, AW","MB, AW","Angle, LC",
                       "Angle, LC","Antrum, AW","Antrum, AW","Antrum, GC",
                       "Antrum, GC","Angle, LC","Angle, LC","UB, LC","UB, LC",
                       "Antrum, LC","Antrum, PW","Antrum, PW",
                       "Angle, LC Antrum, AW","Angle, LC Antrum, AW","LB, GC","LB, GC",NA,NA,
                       NA,NA,NA) |> as.factor(),
            pylori = c("1","1","1","1","0","0",
                       "1","1","1","1","Equivocal","Equivocal","1","1",
                       "1","1","1","1","1","1","1","1","1","1","1","1",
                       "1","1","0","0","Equivocal","Equivocal","1","1",
                       "1","1","1","1","1","1","0","1","1","1","1",
                       "1","1","1","1","1","1","1") |> as.factor(),
               msi = c("MSS","MSS","MSS","MSS",
                       "MSS","MSS","MSS","MSS","MSS","MSS","MSS","MSS",
                       "MSS","MSS","MSS","MSS","MSS","MSS","MSS","MSS","MSS",
                       "MSS","MSS","MSS","MSS","MSS","MSS","MSS","MSS",
                       "MSS","MSS","MSS","MSS","MSS","MSS","MSS","MSS",
                       "MSS","MSS","MSS","MSS","MSS","MSS","MSS","MSS",
                       "MSS","MSS",NA,NA,NA,NA,NA) |> as.factor(),
               ebv = c(NA,NA,0L,0L,NA,NA,NA,NA,
                       NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,0L,0L,
                       0L,0L,NA,NA,0L,0L,0L,0L,NA,NA,NA,NA,NA,NA,
                       NA,NA,NA,NA,NA,NA,NA,NA,NA,1L,1L,NA,NA,NA,NA,
                       NA) |> as.factor(),
              site = c("A", "B", "A", "B", "A", "B", "A", "B", "A", "B", "A", 
                       "B", "A", "B", "A", "B", "A", "B", "A", "B", "A", "B", 
                       "A", "B", "A", "B", "A", "B", "A", "B", "A", "B", "A", 
                       "B", "A", "B", "A", "B", "A", "B", "A", "A", "B", "A", 
                       "B", "A", "B", "A", "A", "A", "A", "A") |> as.factor(),
             cells = c(80L,91L,553L,679L,260L,
                       409L,236L,244L,427L,208L,1111L,383L,275L,111L,202L,
                       389L,580L,216L,154L,208L,170L,162L,419L,544L,
                       317L,155L,148L,274L,442L,269L,84L,331L,82L,118L,
                       401L,203L,83L,163L,173L,39L,110L,463L,363L,162L,
                       43L,163L,325L,253L,479L,2032L,163L,1886L)
  ) |> mutate(sample = paste0("Pat", patient, "-", site)) |> dplyr::select(-patient, -site)

# # In wider format with separate columns for tumor and adjacent cells
# patient_data |>  pivot_wider(names_from = "site", values_from = "cells") |> dplyr::rename(tumor = A, adjacent = B)

# # append patient metadata columns to the seurat object
# # seurat_integrated2 <- as.Seurat(seurat.sce)
# # saveRDS(seurat_integrated2, file="./output/seurat_integrated2.rds")
# 
# seurat_integrated2 <- readRDS("~/workshop/output/seurat_integrated2.rds")
# 
# seurat_integrated2@meta.data <- seurat_integrated2@meta.data |> 
#   mutate(
#     patient = Sample |> str_match("Pat(\\d{2})-([:upper:])") |> as.data.frame() |> pull(2) |> as.factor(),
#     site   = Sample |> str_match("Pat(\\d{2})-([:upper:])") |> as.data.frame() |> pull(3) |> as.factor(),
#     sample = Sample |> str_extract("Pat\\d{2}-[:upper:]"))
# 
# #left_join() removes rownames hence the following three steps are required to conserve them      
# seurat_integrated2@meta.data <- seurat_integrated2@meta.data |> rownames_to_column()
# seurat_integrated2@meta.data <- left_join(seurat_integrated2@meta.data, patient_data, by = "sample")
# seurat_integrated2@meta.data <- seurat_integrated2@meta.data |> column_to_rownames("rowname")
```

```{r pseudobulk}
# library(org.Hs.eg.db)

# pseudo-bulk workflow 

DefaultAssay(seurat_integrated2)

cts <- AggregateExpression(seurat_integrated2, 
                    group.by = c("SingleR.labels", "sample"),
                    slot = "counts",
                    assays = 'originalexp',
                    return.seurat = FALSE)

cts <- cts$originalexp

cts <- cts |> as.data.frame()

annots <- AnnotationDbi::select(org.Hs.eg.db, keys=rownames(cts),
columns=c("SYMBOL","GENENAME"), keytype="ENSEMBL") |>  
  filter(!is.na(SYMBOL)) |> # remove rows that don't have a symbol
  dplyr::add_count(ENSEMBL) |> filter(n == 1) |>  # remove rows that map to two symbols
  dplyr::select(-n) |>
  dplyr::add_count(SYMBOL) |> filter(n == 1) |>  # remove rows that map to two symbols
  dplyr::select(-n)

cts <- cts |> rownames_to_column() |> dplyr::rename(ENSEMBL = "rowname") |> right_join(annots)

rownames(cts) <- cts |> pull(SYMBOL)

cts <- cts |> dplyr::select(-c(ENSEMBL, GENENAME, SYMBOL))

# transpose
cts.t <- t(cts)

# convert to data.frame
cts.t <- as.data.frame(cts.t)

# get values where to split
splitRows <- gsub('_.*', '', rownames(cts.t))

# split data.frame
cts.split <- split.data.frame(cts.t,
                 f = factor(splitRows))

# fix colnames and transpose

cts.split.modified <- lapply(cts.split, function(x){
  rownames(x) <- gsub('.*_(.*)', '\\1', rownames(x))
  t(x)
  
})

# saveRDS(cts.split.modified, file="./output/cts.split.modified.rds")
```

```{r de_cancer}
cts.split.modified <- readRDS("./output/cts.split.modified.rds")

# Let's run DE analysis
# 1. Get counts matrix
cell_type <- vector("list", length = length(cts.split.modified))

for (i in seq_along(cell_type))
  {
  # generate sample level metadata
  colData <- data.frame(sample = colnames(cts.split.modified[[i]]) 
                        ) |> mutate(
    patient = sample |> str_match("Pat(\\d{2})-([:upper:])") |> as.data.frame() |> pull(2) |> as.factor(),
    site   = sample |> str_match("Pat(\\d{2})-([:upper:])") |> as.data.frame() |> pull(3) |> as.factor(),
    site = case_when(site == "A" ~ "non.cancer",
      TRUE ~ "cancer")) |> left_join(patient_data) %>% column_to_rownames(var = 'sample')
  
  cell_type[[i]] <- DESeqDataSetFromMatrix(countData = cts.split.modified[[i]],
                                            colData = colData,
                                            design = ~ patient + site
                                            )
  # run DESeq2
  cell_type[[i]] <- DESeq(cell_type[[i]])

  # Generate results object
  cell_type[[i]] <- results(cell_type[[i]], name = "site_non.cancer_vs_cancer")
  cell_type[[i]] <- as.data.frame(cell_type[[i]])
  }

mirs <- data.frame(
  symbol = c("hsa-miR-106a-5p", "hsa-miR-106b-5p", "hsa-miR-16-1-3p",
             "hsa-miR-16-5p","hsa-miR-15a-5p", "hsa-miR-15b-5p",
             "hsa-miR-1227-3p","hsa-miR-142-5p", "hsa-miR-17-5p",
             "hsa-miR-188-3p","hsa-miR-200b-3p", "hsa-miR-200c-3p",
             "hsa-miR-222-5p","hsa-miR-27a-3p","hsa-miR-30d-3p",
             "hsa-miR-31-5p","hsa-miR-3194-5p","hsa-miR-324-5p",
             "hsa-miR-331-3p","hsa-miR-34a-5p","hsa-miR-34b-5p",
             "hsa-miR-3614-3p","hsa-miR-3616-3p","hsa-miR-369-3p",
             "hsa-miR-374a-5p","hsa-miR-3934-5p","hsa-miR-424-5p",
             "hsa-miR-429","hsa-miR-495-3p","hsa-miR-497-5p",
             "hsa-miR-501-5p","hsa-miR-556-5p","hsa-miR-577",
             "hsa-miR-628-3p","hsa-miR-671-5p","hsa-miR-7-1-3p",
             "hsa-miR-7-2-3p","hsa-miR-877-5p","hsa-miR-93-3p",
             "hsa-miR-940", "hsa-miR-96-5p")
  ) |> mutate(
  symbol = str_replace(symbol, "(-\\dp)?$", "HG"), # replaces -3p & -5p with HG 
  symbol = str_replace(symbol, "hsa-(miR)-", "\\1"), # replace hsa-miR- with miR
  symbol = str_to_upper(symbol) # converts to upper case
               )

common_mirs <- cts.split.modified[[1]] |> as.data.frame() |> rownames_to_column() |> dplyr::select(rowname) |> filter(str_detect(rowname, "MIR")) |> inner_join(mirs, by = c("rowname" = "symbol"))

names(cell_type) <- cts.split.modified |> names()

de_cancer <- map_dfr(
  .x = cell_type,
  ~rownames_to_column(.x) |> filter(rowname %in% c("NOD1", common_mirs)), 
  .id ="cell_type"
  ) 

# saveRDS(de_cancer, file="./output/de_cancer.rds")

de_cancer <- readRDS("~/workshop/output/results_cancer.rds")

de_cancer |> 
  dplyr::rename(
    gene = rowname,
    cell = cell_type,
    baseline = baseMean,
    log2fc= log2FoldChange,
    SE = lfcSE
    ) |>  
  DT::datatable() |> 
  DT::formatRound(digits =2, columns = 3:8)
```

```{r de_pylori}
cts.split.modified <- readRDS("./output/cts.split.modified.rds")

# Let's run DE analysis
# 1. Get counts matrix
cell_type <- vector("list", length = length(cts.split.modified))

for (i in seq_along(cell_type))
  {
  # generate sample level metadata
  colData <- data.frame(sample = colnames(cts.split.modified[[i]]) 
                        ) |> mutate(
    patient = sample |> str_match("Pat(\\d{2})-([:upper:])") |> as.data.frame() |> pull(2) |> as.factor(),
    site   = sample |> str_match("Pat(\\d{2})-([:upper:])") |> as.data.frame() |> pull(3) |> as.factor(),
    site = case_when(site == "A" ~ "non.cancer",
      TRUE ~ "cancer")) |> left_join(patient_data) %>% column_to_rownames(var = 'sample')
  
  cell_type[[i]] <- DESeqDataSetFromMatrix(countData = cts.split.modified[[i]],
                                            colData = colData,
                                            design = ~ pylori
                                            )
  # run DESeq2
  cell_type[[i]] <- DESeq(cell_type[[i]])

  # Generate results object
  cell_type[[i]] <- results(cell_type[[i]], name = "pylori_1_vs_0")
  cell_type[[i]] <- as.data.frame(cell_type[[i]])
  }

names(cell_type) <- cts.split.modified |> names()

de_pylori <- map_dfr(
  .x = cell_type,
  ~rownames_to_column(.x) |> filter(rowname %in% c("NOD1", common_mirs)), 
  .id ="cell_type"
  ) 

# saveRDS(de_pylori, file="./output/de_pylori.rds")

de_pylori <- readRDS("~/workshop/output/de_pylori.rds")

de_pylori |> 
  dplyr::rename(
    gene = rowname,
    cell = cell_type,
    baseline = baseMean,
    log2fc= log2FoldChange,
    SE = lfcSE
    ) |>  
  DT::datatable() |> 
  DT::formatRound(digits =2, columns = 3:8)
```

```{r expression_plot}
map_dfr(
.x = cts.split.modified,
~ as.data.frame(.x) |>
rownames_to_column() |>
filter(rowname == "PTPRC") |>
column_to_rownames(var = "rowname") |>
pivot_longer(cols = everything(), names_to = "rowname", values_to = "count") |>
full_join(colData |>  rownames_to_column(), by = "rowname" ),
.id = "label") |>
ggplot(aes(x = site, y = count)) +
geom_boxplot() +
geom_point(shape = 1, position = position_jitter(width = 0.1),
alpha = 0.20) +
facet_grid(. ~label) +
labs(title = "Boxplots of CD45 expression in hematopoietic cell types",
y = "CD45 transcript count in patients (n > 20)", x = "Biopsy site")
```

# Normalization
Here, we'll do a straightforward "naive" normalization approach: log-transformation and library size scaling

A few great alternatives have been propose (eg. SCTransform, which is also built into the Seurat package), but their use is a bit more nuanced, so I'll avoid them for this workshop. I'd recommend checking them out though.

```{r}
seurat <- NormalizeData(seurat)
```

# Dimensionality Reduction
Usually from here, we're interested in doing some exploratory analysis of our data. We want to make visualizations like tSNE/UMAP plots to see the "structure" of our data. We want to cluster cells based on their expression profiles. With that, we can start doing things like differential expression to identify which cell types are present in the data and which genes are good markers to distinguish different cellular populations.

All of this analysis requires us to perform "Dimensionality Reduction". Currently in our data, each cell could be represented as a single point in 18,340-dimensional space (each dimension corresponding to a single gene). Obviously, this is impossible for us to manually explore unless we have prior knowledge about important dimensions/genes for our system. Luckily, gene expression is structured. If two genes are perfectly correlated, we could represent two genes worth of information with a single value. The goal of dimensionality reduction is to take advantage of this structure to "refine" 18,340 dimensions worth of information into a smaller number of dimensions.

## Finding highly variable genes
Some of the genes in our data are like housekeeping genes and their expression is pretty constant across different cell types/states. As such, these genes don't really contribute to the structure/heterogeneity of our data. One step we can take to refine our data down is to ignore these genes and only focus on genes that do contribute. To do this, we measure the variance of each gene's expression across the entire population of cells. Genes with low variance are things like housekeeping genes and don't tell us much. Cell type-specific markers are examples of genes that would have high variance because in some cells there is little-to-no expression, while in others they could have very high expression. 

By focusing only on genes with high variance, we can retain much of the structure of our data in as few as a couple thousand genes.

Let's find the top 2000 variable genes:

```{r}
seurat <- FindVariableFeatures(seurat,
                               selection.method="vst",
                               nfeatures = 2000)
```

And just for our interest, let's check out the top 20 variable genes
```{r}
head(VariableFeatures(seurat), n=20)
```

From this alone, we see many immune signatures we may expect, like key markers of MHC components, immunoglobulin components, etc.

## Principal Component Analysis (PCA)
So 2000 dimensions is better than 18,000 but it's still too large to explore. Luckily we have more tools to refine the structure down into fewer dimensions.

As I mentioned before, gene expression is quite modular due to co-regulation. PCA can identify this structure and generates **principal components (PCs)**, which are single dimensions/axes/features that are a linear combination of gene expression measurements (see PPT slides). So instead of having a genes-by-cells matrix, we will *embed* our data in *PC space* and have PC-by-cell matrix.

### Optional Demo: Correlation heatmap of variable genes
To demonstrate the structure of gene expression, we can simply calculate the pearson correlation of these top 2000 variable genes.
```{r}
exp <- as.matrix(seurat[["RNA"]]@data[VariableFeatures(seurat),])
exp <- t(cor(t(exp), method="pearson"))
```

```{r}
pheatmap::pheatmap(exp,
                   breaks = seq(-0.2, 0.2, length.out=101),
                   clustering_method = "ward.D2",
                   show_rownames = F,
                   show_colnames = F)
```

You'll see that there are distinct modules of highly correlated genes. These often correspond to programs of distinct cell types.

### PCA continued
Because PCA works by identifying sources of variation in the data, we have to standardize the means and variance of our data using a Z-score transformation. This is important so that every gene is considered equally and genes with a higher average expression don't get weighed more in the PCA. Seurat has a convenient function for this:

```{r}
seurat <- ScaleData(seurat)
```

Side Note: Sometimes, there are technical "artifacts" in our embeddings that we don't want. Eg. Sometimes on PCA, you'll find that certain PCs are directly correlated with things like percent.mito (cell viability), cell cycle, etc. The effects of these variables can be "normalized" out through a linear regression model (we often call it "regressing it out"). I'm not going to add it here, but you could add the vars.to.regress option to ScaleData() to do this. eg. ScaleData(seurat, vars.to.regress=c("percent.mito", "S.Score", "G2M.Score)).

Now we can run the PCA

```{r}
seurat <- RunPCA(seurat)
```

By default, Seurat's RunPCA() function will print out the genes with the top "loadings" for each PC.

Genes with the highest positive loading will be the "strongest forces" contributing to a cell's position higher up on that PC. Genes with the the lowest (negative) loading do the opposite, lowering a cell's value on that PC.

Let's look into that a big more

### Visualize the PCA
Seurat has a couple easy visualization functions. DimPlot() lets us look at our embeddings, colouring the points by any categorical info stored in the Seurat object (eg. cluster) and FeaturePlot() lets us look at our embeddings, colouring points by any quantitative value in our Seurat object (gene expression, quantitative metadata, etc). 

```{r}
DimPlot(seurat, reduction='pca')
```

And quickly, let's look at a couple of those genes with high loadings with PC1/2. With FeaturePlot() you can give it as many genes as you want and it will generate multiple panels
```{r}
FeaturePlot(seurat, features=c("CD3D", "LYZ", "NKG7", "CD79A"),
            cols = c('lightgrey', 'red'))
```

From this alone, we can see pretty distinct populations of monocytes (LYZ) and B cells (CD79A), but interestingly, we don't get much separation of our T cells (CD3D) and NK cells (NKG7).

This is an important feature of PCA: it is a *linear* dimensionality reduction (feel free to read up on this more). For complex data, it will be impossible to capture all of the data's structure in as few as two dimensions.

Let's look at how much variation there is along each of the 50 PCs we computed.

```{r}
ElbowPlot(seurat, ndims=50)
```

You can see that eventually the variance per PC plateaus. This is usually the point were the PCA is just capturing noise in the data. This means that ~20-30PCs have more variance above this noise, suggesting that these are probably the PCs that capture biological information.

So, we went from 18,000 dimensions, to 2000 dimensions, and now down to 20-30. We're getting there, but unfortunately 30 dimensions are still hard to visualize.

## UMAP: non-linear dimensionality reduction
You've probably heard of tSNE (t-distributed stochastic neighbor embedding) or UMAP (uniform manifold approximation and projection). These are both examples of non-linear dimensionality reduction methods. They are capable of embedding complex data in a very small number of dimensions. Both essentially try to optimize 2/3-dimensional organization of data where the distance between points in low dimension space is similar to distances in high dimensional space. As a result, similar data points in gene expression will be placed close to each other in the 2-3D embedding. In the end, the axes themselves **mean nothing** (eg. high x-axis values do not mean higher expression of certain genes). Similarly, because of how the algorithms are set up, things like the shape of the "blobs" on the plot or the relative positions of the "blobs" are also meaningless. 

I'd recommend checking out this article on tSNE to see how different parameters and input data can affect the resultant tSNE embedding: https://distill.pub/2016/misread-tsne/

All this said, they are still incredibly effective tools for visualizing our data. 

It's common to use PCA embeddings as the input for the tSNE/UMAP algorithm. This has two main benefits: 1) compute time--these algorithms can be very slow, and even 2000 genes can take a while to run. 2) The idea of PCA was pull out the signal from our data in a relatively small number of PCs. By inputting only the "meaningful" PCs (ie. those above that background amount of noise we talked about above) into the UMAP/tSNE algorithm, we're helping the signal-to-noise of the data for these tools to work with. From that "elbow plot" above, we saw that the first 20-30 PCs capture most of our biological variation. Let's meet in the middle and use 25 PCs for the UMAP algorithm


```{r}
seurat <- RunUMAP(seurat, dims=1:30)
```

### Visualize the UMAP
```{r}
DimPlot(seurat) #UMAP will have been made the default visualization for DimPlot by Seurat
```
 
 Great! We're now getting really good seperation of populations. Let's look at those same markers.
 
```{r}
FeaturePlot(seurat, features=c("CD3D", "LYZ", "NKG7", "CD79A"),
            cols = c('lightgrey', 'red'))
```

Notice that we now have the complete separation of a population of NKG7+ cells that were completely obscured in the PCA plots.

# Save point
This is a good time to save a version of processed data. Many steps after this point can be modified, so it's nice to have this version saved for you to come back to.

```{r}
saveRDS(seurat, file="../output/pbmc_seurat_filtered.rds")
```

